File: shared/src/commonMain/kotlin/eu/vitamoments/app/data/models/dto/user/FriendshipDto.kt
package eu.vitamoments.app.data.models.dto.user

import eu.vitamoments.app.data.enums.FriendshipDirection
import eu.vitamoments.app.data.enums.FriendshipStatus
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.uuid.Uuid

@Serializable
sealed interface FriendshipDto{
    val uuid: Uuid
    val status: FriendshipStatus
    val createdAt: Long
    val updatedAt: Long
}

@Serializable
@SerialName("PENDING")
data class PendingFriendshipDto(
    override val uuid: Uuid,
    override val createdAt: Long,
    override val updatedAt: Long,
    val direction: FriendshipDirection,
    val friend: PublicUserDto,
) : FriendshipDto {
    override val status: FriendshipStatus = FriendshipStatus.PENDING
}

@Serializable
@SerialName("ACCEPTED")
data class AcceptedFriendshipDto(
    override val uuid: Uuid,
    override val createdAt: Long,
    override val updatedAt: Long,
    val friend: PrivateUserDto,
) : FriendshipDto {
    override val status: FriendshipStatus = FriendshipStatus.ACCEPTED
}

File: webApp/src/api/Oldtypes/friend/friendshipDto.ts
import type { FriendshipStatus } from "./friendshipStatus";
import type { FriendshipDirection } from "./friendshipDirection";

export type FriendshipDto = PendingFriendshipDto | AcceptedFriendshipDto;

interface BaseFriendshipDto {
  uuid: string;
  status: FriendshipStatus; // redundant maar ok
  createdAt: number;
  updatedAt: number;
  otherUserId: string; // jouw otherUserId in backend
}

export interface PendingFriendshipDto extends BaseFriendshipDto {
  type: "PENDING";
  direction: FriendshipDirection; // INCOMING | OUTGOING
}

export interface AcceptedFriendshipDto extends BaseFriendshipDto {
  type: "ACCEPTED";
  acceptedAt: number;
}
